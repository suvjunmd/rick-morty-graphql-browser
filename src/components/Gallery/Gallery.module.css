.container {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
  gap: 1rem;
}

.card {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 0.5rem;
}

.statusAlive {
  background: linear-gradient(0deg, #388e3c, #4caf50);
}

.statusDead {
  background: linear-gradient(0deg, #d32f2f, #f44336);
}

.statusUnknown {
  background: linear-gradient(0deg, #1976d2, #2196f3);
}

.image {
  width: 100%;
  object-fit: cover;
}

.link {
  color: black;
  font-size: 20px;
}
