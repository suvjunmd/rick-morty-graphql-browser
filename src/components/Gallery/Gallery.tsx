import { Link } from "react-router-dom";
import { useQuery } from "@apollo/client";
import Pagination from "../Pagination";
import { FilterConfig } from "../../utils";
import styles from "./Gallery.module.css";
import { graphql } from "../../gql";

interface GalleryProps {
  filterConfig: FilterConfig;
  page: number;
  onNavigateToPage: (page: number) => void;
}

export const GET_CHARACTERS_QUERY = graphql(`
  query GetCharacters($page: Int, $filter: FilterCharacter) {
    characters(page: $page, filter: $filter) {
      info {
        pages
      }
      results {
        id
        name
        status
        image
      }
    }
  }
`);

export default function Gallery({
  filterConfig,
  page,
  onNavigateToPage,
}: GalleryProps) {
  const { loading, error, data } = useQuery(GET_CHARACTERS_QUERY, {
    variables: { page, filter: filterConfig },
  });

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error : {error.message}</p>;
  if (!data || !data.characters) return <p>No results</p>;

  return (
    <>
      <div className={styles.container}>
        {data.characters.results?.map((character) => (
          <div
            className={`${styles.card} ${
              character?.status === "Alive" ? styles.statusAlive : ""
            }${character?.status === "Dead" ? styles.statusDead : ""}${
              character?.status === "unknown" ? styles.statusUnknown : ""
            }`}
            key={character?.id}
          >
            <p>
              <Link to={`/characters/${character?.id}`} className={styles.link}>
                {character?.id} - {character?.name}
              </Link>
            </p>
            <img
              width="200"
              height="200"
              alt=""
              src={character?.image ?? undefined}
              className={styles.image}
            />
          </div>
        ))}
      </div>
      <Pagination
        page={page}
        totalPages={data.characters.info?.pages ?? 1}
        onNavigateToPage={onNavigateToPage}
      />
    </>
  );
}
