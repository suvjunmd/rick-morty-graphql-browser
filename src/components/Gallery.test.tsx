import { expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import Gallery, { GET_CHARACTERS_QUERY } from "./Gallery";
import { MemoryRouter } from "react-router-dom";

const mocks = [
  {
    request: {
      query: GET_CHARACTERS_QUERY,
      variables: {
        page: 1,
        filter: {},
      },
    },
    result: {
      data: {
        characters: {
          info: {
            pages: 1,
          },
          results: [
            {
              id: "1",
              name: "Rick Sanchez",
              status: "Alive",
              image: "1.jpeg",
            },
            {
              id: "2",
              name: "Morty Smith",
              status: "Dead",
              image: "2.jpeg",
            },
            {
              id: "3",
              name: "Summer Smith",
              status: "unknown",
              image: "3.jpeg",
            },
          ],
        },
      },
    },
  },
];

it("should render gallery", async () => {
  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <MemoryRouter>
        <Gallery filterConfig={{}} page={1} onNavigateToPage={() => {}} />
      </MemoryRouter>
    </MockedProvider>
  );
  expect(await screen.findByText("Loading...")).toBeInTheDocument();
  expect(await screen.findByText("2 - Morty Smith")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

it("should render no results message", async () => {
  const noResultsMock = {
    request: {
      query: GET_CHARACTERS_QUERY,
      variables: {
        page: 1,
        filter: {},
      },
    },
    result: {},
  };
  render(
    <MockedProvider mocks={[noResultsMock]} addTypename={false}>
      <Gallery filterConfig={{}} page={1} onNavigateToPage={() => {}} />
    </MockedProvider>
  );
  expect(await screen.findByText("No results")).toBeInTheDocument();
});

it("should render network error message", async () => {
  const errorMock = {
    request: {
      query: GET_CHARACTERS_QUERY,
      variables: {
        page: 1,
        filter: {},
      },
    },
    error: new Error("An error occurred"),
  };
  render(
    <MockedProvider mocks={[errorMock]} addTypename={false}>
      <Gallery filterConfig={{}} page={1} onNavigateToPage={() => {}} />
    </MockedProvider>
  );
  expect(
    await screen.findByText("Error : An error occurred")
  ).toBeInTheDocument();
});
