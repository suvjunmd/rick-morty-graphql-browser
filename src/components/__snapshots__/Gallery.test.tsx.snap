// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render gallery 1`] = `
<div>
  <div
    class="gallery"
  >
    <div
      class="gallery-item"
    >
      <p>
        <a
          href="/characters/1"
        >
          1
           - 
          Rick Sanchez
        </a>
        <span
          class="alive"
        />
      </p>
      <img
        alt=""
        height="200"
        src="1.jpeg"
        width="200"
      />
    </div>
    <div
      class="gallery-item"
    >
      <p>
        <a
          href="/characters/2"
        >
          2
           - 
          Morty Smith
        </a>
        <span
          class="dead"
        />
      </p>
      <img
        alt=""
        height="200"
        src="2.jpeg"
        width="200"
      />
    </div>
    <div
      class="gallery-item"
    >
      <p>
        <a
          href="/characters/3"
        >
          3
           - 
          Summer Smith
        </a>
        <span
          class="unknown"
        />
      </p>
      <img
        alt=""
        height="200"
        src="3.jpeg"
        width="200"
      />
    </div>
  </div>
  <div
    class="pagination"
  >
    <button
      disabled=""
    >
      Previous
    </button>
    <div
      class="page-number"
    >
      page 
      1
       of 
      1
    </div>
    <button
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;
